### PR TITLE
added Timer is_time_left()

### PR DESCRIPTION
**Pull request by AlexHolly**
part of #7928

TODO docs

### CODE REVIEW

**Comment by neikeq** [comment]
I think `is_running` would be a better name.

File: scene/main/timer.cpp [hunk 1]
```diff
@@ -116,6 +116,10 @@ void Timer::stop() {
 	autostart = false;
 }
 
+bool Timer::is_time_left() const {
+	return get_time_left() > 0;
```
Comment (bojidar-bg) [inline scene/main/timer.cpp]: I'm not sure if this would work properly, since it is floating-point comparison without epsilons...
Maybe `get_time_left() > 0.0001` would be more stable?

**Review by bojidar-bg — APPROVED** [review]
You didn't run doctool tho, but that's not a big issue (next documenter would)

File: scene/main/timer.cpp (hunk 1 above)
Comment (akien-mga) [inline scene/main/timer.cpp]: Check line 137... :)

Comment (bojidar-bg) [inline scene/main/timer.cpp]: Ok, I retract my claims then :smiley:

**Comment by AlexHolly** [comment]
I don't think that is_running is specific enough and since there already is one confusing function(is_active) it would make things even worse. is_running would return true even when is_active is false that means it's not active but still running doesn't sounds good to me. Since is_time_left is using time_left I think it's better to understand. Let's Vote :D

:+1:  : is_running
:-1:  is_time_left

**Comment by neikeq** [comment]
Then you could change it to `return is_active() && get_time_left() > 0`.
As far as I understand, the issue was that there is no intuitive way to know if a timer is still running. Well, `is_time_left` (btw, should be `has_time_left`) won't solve that, and I don't think it's worth adding just as a shortcut for `get_time_left() > 0`, unless it has some other value rather than just being a shortcut.

**Comment by AlexHolly** [comment]
The main issue is that is_active is called is_active and not is_paused I was looking for way to check if the timer is already running so is_active looked very promising. is(has)_time_left will make thinks a bit better but won't solve the real issue is_active is not specific enough.

**Comment by neikeq** [comment]
**TL;DR:**
- Fix the `get_time_left` documentation.
- Rename `active` to `paused` and inverting the boolean value.
- Fix a potential erroneous state when `start` is called while `active` is false.
- Propose `stopped` as an intuitive method to check if the timer stopped.

-----

_Note that when I say "active" I refer to the `active` variable in the Timer class._

After reading the documentation I found a few issues:

- `get_time_left` returns 0 if the timer stopped (or never started), otherwise it returns the time left for timeout in seconds _even if the timer is not active!!_. **There is a documentation problem here since the method description suggests that it would also return 0 if the started but is not active, which is wrong.** IMO `active` should be renamed to `paused` inverting the boolean value.
- There is a potential bug with `start` which changes processing to true even if `active` was set to false previously. That could result in a timer that is processing while `is_active` would return false. Either `start` should not enable processing in such cases, or it should set `active` to true.
- Finally, we could add a method named `stopped`. It would check `time_left == -1` (`stop` sets `time_left` to -1), not the same as `get_time_left()` which returns 0 if `time_left` is -1.

**Comment by AlexHolly** [comment]
```
extends Node2D

func _ready():
	
	var timer = Timer.new()
	add_child(timer)
	timer.set_one_shot(true)
	timer.set_name("Timer")
	
	print("time_left: " + str($Timer.get_time_left()))
	print("is_paused: " + str($Timer.is_paused()))
	print("is_running: " + str($Timer.is_running()))
	print("START")
	
	$Timer.start()
	
	print("time_left: " + str($Timer.get_time_left()))
	print("is_running: " + str($Timer.is_running()))
	print("is_paused: " + str($Timer.is_paused()))
	
	print("PROCESS")
	set_process(true)
	
func _process(delta):
	
	
	if $Timer.is_running():
		print("time_left: " + str($Timer.get_time_left()))
		
		if $Timer.get_time_left()<0.5 && !$Timer.is_paused():
			print("PAUSE")
			$Timer.set_paused(true)
		
	
```

File: scene/main/timer.cpp [hunk 2]
```diff
@@ -116,16 +116,20 @@ void Timer::stop() {
...
 }
 
-bool Timer::is_active() const {
-	return active;
+bool Timer::is_paused() const {
+	return paused;
+}
+
+bool Timer::is_running() const {
+	return get_time_left() > 0;
```
Comment (neikeq) [inline scene/main/timer.cpp]: Although, apparently the time left can never be 0 while the timer is running, I would do `time_left != -1` instead. Also, could `is_running` be confusing? It may return true even if the timer is paused, so maybe `stopped` is a better name.
Just my opinion. PR looks good.

**Comment by neikeq** [comment]
`doc/base/classes.xml` has merged conflicts. Try `git add --patch doc/base/classes.xml` to select just the part you mean to commit.

File: scene/main/timer.cpp (hunk 2 above)
Comment (bojidar-bg) [inline scene/main/timer.cpp]: @neikeq I disagree about `!= -1`. One day a terrible platform might come out, in which `-1.0f != -1.0f`, and then we are all going to have problems...

Comment (neikeq) [inline scene/main/timer.cpp]: Ah, right. It's floating point. Forget it.

**Comment by AlexHolly** [comment]
There are too many changes in the docs. I will open a new PR for the docs.

**Comment by akien-mga** [comment]
Note that the docs are a bit in a limbo right now, due to the 3.0 change to expose properties, we need to be cautious when updating the template to make sure we don't lose information. So best leave it alone for now if it seems hairy :)

**Review by neikeq — APPROVED** [review]
